File: _pattern/singleton/rpt.cpp
/*
싱글톤 클래스를 기반으로 한다. 
싱글톤 클래스에서 인스턴스 생성 시 동적으로 전달인자를 설정하고,
생성자에 해당 값을 전달하여 total의 초기값으로 설정하도록 한다. 
total의 값을 출력하라
*/
#include <iostream>
using namespace std;


class Sgt {
private:
    Sgt(){}
    static int total;
    static Sgt* pist;
public:
    ~Sgt(){}
    static Sgt* getInstance(int value = 10){
        if(pist == NULL){
            pist = new Sgt();
            total = value;
        }
        
        return pist;
    }
    void setTotal(int num) {
        total = num;
    }
    int getTotal(){
        return total;
    }
};

Sgt* Sgt::pist = NULL;
int Sgt::total = 0; // 정적 멤버는 클래스 외부에서 한 번 반드시 정의(내부 선언만 가능)

int main () {
    Sgt* ist1 = Sgt::getInstance(20);
    Sgt* ist2 = Sgt::getInstance(40); // 선언으로 값 변경 불가
    Sgt* ist3 = Sgt::getInstance(55);

    cout << "생성자 값" << endl;
    cout << ist1 -> getTotal() << endl;    
    cout << ist2 -> getTotal() << endl;
    cout << ist3 -> getTotal() << endl;

    cout << "멤버함수 포인터 간접접근" << endl;
    ist1 -> setTotal(30);
    cout << ist1 -> getTotal() << endl;    
    cout << ist2 -> getTotal() << endl;
    cout << ist3 -> getTotal() << endl;    


    return 0;
}

File: pattern/singleton/rpt.js
/**
 * C++에서 작성한 싱글톤 패턴 코드를 자바스크립트 코드로 변환하라. 

문법 체크포인트)
1. 자스에서는 생성자 이름을  constructor()로 사용함.
2. :: 연산자 대신 . 연산자를 사용
3. 전달인자에 타입 지정하지 않음.
 */
class Sgt {
    static ist = null;
    static total = 0;

    constructor() {
        if(Sgt.ist) {
            console.error("생성자 신규 불가!");
            return Sgt.ist;
        }
    }

    static getIncetance(value = 10) {
        if(Sgt.ist === null) {
            Sgt.ist = new Sgt();
            Sgt.total = value;
        }
        return Sgt.ist
    }
    
    setTotal(num){
        Sgt.total = num;
    }
    getTotal(){
        return Sgt.total
    }
}

// 출력 및 멤버함수 호출
const ist1 = Sgt.getIncetance(20);
const ist2 = Sgt.getIncetance(30);
const ist3 = Sgt.getIncetance(40);

console.log("생성자로 초기값 수정 시 => 초기값만 적용되며, 새로운 인스턴스 생성 불가(값 적용X)");
console.log(ist1.getTotal());
console.log(ist2.getTotal());
console.log(ist3.getTotal());

console.log("멤버함수 값 변경 모든 인스턴스 값 변경")
ist1.setTotal(100);
console.log(ist1.getTotal());
console.log(ist2.getTotal());
console.log(ist3.getTotal());